pysindy/optimizers/trapping_sr3.py

```diff
@@ -0,0 +1,510 @@
+        self.A_history_ = []
+        self.m_history_ = []
+        self.PW_history_ = []
+        self.PWeigs_history_ = []
+        self.history_ = []
+        self.objective_history = objective_history
+        self.unbias = False
+        self.use_constraints = (constraint_lhs is not None) and (
+            constraint_rhs is not None
+        )
+
+        if self.use_constraints:
+            if constraint_order not in ("feature", "target"):
+                raise ValueError(
+                    "constraint_order must be either 'feature' or 'target'"
+                )
+
+            self.constraint_lhs = constraint_lhs
+            self.constraint_rhs = constraint_rhs
+            self.unbias = False
+            self.constraint_order = constraint_order
+
+    def _update_coef_constraints(self, H, x_transpose_y, P_transpose_A):
+        g = x_transpose_y + P_transpose_A / self.eta
+        inv1 = np.linalg.pinv(H, rcond=1e-15)
+        inv2 = np.linalg.pinv(
+            self.constraint_lhs.dot(inv1).dot(self.constraint_lhs.T), rcond=1e-15
+        )
+
+        rhs = g.flatten() + self.constraint_lhs.T.dot(inv2).dot(
+            self.constraint_rhs - self.constraint_lhs.dot(inv1).dot(g.flatten())
+        )
+        rhs = rhs.reshape(g.shape)
+        return inv1.dot(rhs)
+
+    def _update_A(self, A_old, PW):
+        """Update the symmetrized A matrix"""
+        eigvals, eigvecs = np.linalg.eigh(A_old)
+        eigPW, eigvecsPW = np.linalg.eigh(PW)
+        r = A_old.shape[0]
+        A = np.diag(eigvals)
+        for i in range(r):
+            if eigvals[i] > self.gamma:
+                A[i, i] = self.gamma
+        return eigvecsPW @ A @ np.linalg.inv(eigvecsPW)
+
+    def _convergence_criterion(self):
+        """Calculate the convergence criterion for the optimization over w"""
+        this_coef = self.history_[-1]
+        if len(self.history_) > 1:
+            last_coef = self.history_[-2]
+        else:
+            last_coef = np.zeros_like(this_coef)
+        err_coef = np.sqrt(np.sum((this_coef - last_coef) ** 2))
+        return err_coef
+
+    def _m_convergence_criterion(self):
+        """Calculate the convergence criterion for the optimization over m"""
+        return np.sum(np.abs(self.m_history_[-2] - self.m_history_[-1]))
+
+    def _objective(self, x, y, coef_sparse, A, PW, q):
+        """Objective function"""
+
+        # Compute the errors
+        R2 = (y - np.dot(x, coef_sparse)) ** 2
+        A2 = (A - PW) ** 2
+        L1 = self.threshold * np.sum(np.abs(coef_sparse.flatten()))
+
+        # convoluted way to print every max_iter / 10 iterations
+        if q % max(int(self.max_iter / 10.0), 1) == 0 or self.threshold != 0.0:
+            row = [q, 0.5 * np.sum(R2), 0.5 * np.sum(A2) / self.eta, L1]
+            print("{0:12d} {1:12.5e} {2:12.5e} {3:12.5e}".format(*row))
+        return 0.5 * np.sum(R2) + 0.5 * np.sum(A2) / self.eta + L1
+
+    def _reduce(self, x, y):
```

Comment:
This method is over 200 lines long. Can we break it up into smaller components? E.g. put the initializations and precomputations into separate helper methods. Or maybe there are parts of the optimization loop that can be moved to helper functions?

Reply:
Yes this can definitely be broken up into helper methods, working on this now